### PR TITLE
Allow all Petsc matrix types to proceed from a common base class

### PR DESCRIPTION
**Pull request by lindsayad**
And critically, always allow retrieving a PetscMatrix context from a PETSc mat. Eventually we may be able to remove the PetscMatrix from Mat constructors entirely. That would be good

This change was strongly suggested by the need to convert LinearSolver to be able to handle StaticCondensation

### CODE REVIEW

**Review by roystgnr — CHANGES_REQUESTED** [review]

File: src/solvers/petsc_linear_solver.C [hunk 1]
```diff
@@ -194,10 +194,10 @@ void PetscLinearSolver<T>::init (const char * name)
...
 
       if (strcmp(ksp_type, "preonly"))
-        {
-          ierr = KSPSetInitialGuessNonzero (_ksp, PETSC_TRUE);
-          LIBMESH_CHKERR(ierr);
-        }
+      {
+        ierr = KSPSetInitialGuessNonzero (_ksp, PETSC_TRUE);
+        LIBMESH_CHKERR(ierr);
+      }
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply src/solvers/petsc_linear_solver.C:200]: I'm down 😆

File: src/solvers/petscdmlibmeshimpl.C [hunk 2]
```diff
@@ -689,27 +689,29 @@ static PetscErrorCode DMlibMeshJacobian(DM dm, Vec x, Mat jac, Mat pc)
   ierr = DMlibMeshGetSystem(dm, _sys); CHKERRQ(ierr);
   NonlinearImplicitSystem & sys = *_sys;
 
-  PetscMatrix<Number> the_pc(pc,sys.comm());
-  PetscMatrix<Number> Jac(jac,sys.comm());
+  libmesh_assert(pc);
+  libmesh_assert(jac);
+  PetscMatrixBase<Number> * the_pc = PetscMatrixBase<Number>::get_context(pc);
+  PetscMatrixBase<Number> * Jac = PetscMatrixBase<Number>::get_context(jac);
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply src/solvers/petscdmlibmeshimpl.C:695]: In the other case in `petsc_nonlinear_solver.C` the two are allowed to be `nullptr`. However, for the PC in `petsc_nonlinear_solver.C` I never allow reseat the pointer, so I added a `const` there

**Review by roystgnr — APPROVED** [review]

**Comment by lindsayad** [comment]
beautiful failed code

**Comment by lindsayad** [comment]
it could be good to have such a macro in the future. For now, I think we have a good change in https://github.com/libMesh/libmesh/pull/3925

**Pull request merged**